File: 36-valid-sudoku/valid-sudoku.py
import math
class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        cols = {}
        rows = {}
        squares = {}

        for r, row in enumerate(board):
            for c, val in enumerate(row):
                if val == ".":
                    continue    
                
                rowSet = rows.get(r)
                if not rowSet:
                    rowSet = set()
                    rows[r] = rowSet
                
                if val in rowSet:
                    return False
                else:
                    rowSet.add(val)

                colSet = cols.get(c)
                if not colSet:
                     colSet = set()
                     cols[c] = colSet

                if val in colSet:
                    return False
                else:
                    colSet.add(val)
                
                squareX = r // 3
                squareY = c // 3
                squareCoordinates = f"{squareX},{squareY}"

                squareSet = squares.get(squareCoordinates)
                if not squareSet:
                    squareSet = set()
                    squares[squareCoordinates] = squareSet
                
                if val in squareSet:
                    return False
                else:
                    squareSet.add(val)

        return True
                    
                    
                    
        

File: 36-valid-sudoku/valid-sudoku.js
/**
 * @param {character[][]} board
 * @return {boolean}
 */
var isValidSudoku = function(board) {
    const rows = {};
    const cols = {};
    const squares = {};

   for(let r = 0; r < board.length; r++) {
        for (let c = 0; c < board[r].length; c++) {
            const value = board[r][c];
            if(value === '.') continue

            let rowSet = rows[r];
            if (!rowSet) {
                rowSet = new Set();
                rows[r] = rowSet;  
            }

            if (rowSet.has(value)) {
                return false;
            } else {
                rowSet.add(value);
            }

            let colSet = cols[c];
            if (!colSet) {
                colSet = new Set();
                cols[c] = colSet; 
            }

            if (colSet.has(value)) {
                return false;
            } else {
                colSet.add(value);
            }

            const squareX = Math.floor(r / 3);
            const squareY = Math.floor(c / 3);
            const squareCoordinates = `${squareX},${squareY}`;
            console.log(squareCoordinates)

            let squareSet = squares[squareCoordinates];
            if (!squareSet) {
                squareSet = new Set();
                squares[squareCoordinates] = squareSet;  
            }

            if (squareSet.has(value)) {
                return false;
            } else {
                squareSet.add(value);
            }
        }
    }

    return true;
};